src/player/SurvivalBlockBreakHandler.php

```diff
@@ -65,8 +67,27 @@ private function calculateBreakProgressPerTick() : float{
 		if(!$this->block->getBreakInfo()->isBreakable()){
 			return 0.0;
 		}
-		//TODO: improve this to take stuff like swimming, ladders, enchanted tools into account, fix wrong tool break time calculations for bad tools (pmmp/PocketMine-MP#211)
-		$breakTimePerTick = $this->block->getBreakInfo()->getBreakTime($this->player->getInventory()->getItemInHand()) * 20;
+		$breakTimePerSecond = $this->block->getBreakInfo()->getBreakTime($this->player->getInventory()->getItemInHand());
+
+		if($this->player->isUnderwater()){ //TODO: Aqua Affinity enchantment
+			$breakTimePerSecond *= 5;
+		}
+		if(!$this->player->onGround){
+			$breakTimePerSecond *= 5;
+		}
+		if($this->player->canClimb() && $this->player->getWorld()->getBlock($this->player->getPosition())->canClimb()){
+			$breakTimePerSecond *= 5;
+		}
+
+		$effects = $this->player->getEffects();
+		if(($miningFatigue = $effects->get(VanillaEffects::MINING_FATIGUE())) !== null){
+			$breakTimePerSecond *= $breakTimePerSecond / ($breakTimePerSecond * 0.2 ** $miningFatigue->getEffectLevel());
+		}
+		if(($haste = $effects->get(VanillaEffects::HASTE())) !== null){
+			$breakTimePerSecond -= $breakTimePerSecond * 0.2 * min($haste->getEffectLevel(), 5);
```

Comment:
Same again here - I'm pretty sure the values differ from Java.

Reply:
Why maximize the haste level to 5? This is normally an effect that can theoretically have a level of up to 255. The only limit is haste 3 for arm animation.

Reply:
Probably because the break time can become negative if the haste level is higher than 5